pkg/app/component.go

```diff
@@ -274,6 +283,11 @@ func (c *Compo) mount(d Dispatcher) error {
 			Tag("kind", c.Kind())
 	}
 
+	if Initialiser, ok := c.self().(Initialiser); ok && c.disp == nil {
```

Comment:
Here `disp` should be equal to nil. There is no need for a check.

Reply:
If I don't check this it will be called multiple times in the client (on every mount). I checked this in a real projekt.

Reply:
If you dismount a component you don't clear the dispatcher field. You only set this to nil. So if it gets mounted again it still has the dispatcher set. The init is about the underlying data in the structure and if you swap a component it will be dismounted and can be reused.

Reply:
That should not happen. 
There is a check on Mounted
```go
func (c *Compo) Mounted() bool {
	return c.dispatcher() != nil &&
		c.ctx != nil &&
		c.ctx.Err() == nil &&
		c.root != nil && c.root.Mounted() &&
		c.self() != nil
}
```

when a component is dismounted, the context is canceled, which sets an error on the context.
If no there is a bug that should be resolved first. Do you have a sample that could help me reproduce it?

Reply:
In this case, the context was canceled and c.self() is nil, so it returns false. I don't understand whats the problem here? After the dismount, it is a "dead" data structure. I only use the "disp" because this way I do not need another field to indicate it was being initialized before. 

You can check this with my switcher [example](https://github.com/metatexx/go-app-pkgs/blob/master/examples/mountpoint/main.go) just add a `func (c *tab) OnInit() { c.Title+="+" }` which adds something to the title string. If you run it without the check for `disp != nil` you will see that on every (re-)mount the string gets added as the init is called again. If you add it, it will just be initialized once and this is what I made this for.

P.S.: I have a `dbg.Log()` (from [app-go-pkgs/dbg](https://github.com/metatexx/go-app-pkgs/blob/master/dbg/dbg.go)) in all of the functions so I can see what it is calling, when and from where.

Reply:
This is added because you are reusing a dismounted component since you keep a reference on it. Everytime your init is called, the title is appended at the end. You would not have this behavior if you just assign it. Also your implementation would me more safe if you where using a function that return a fresh created version of the component.

Reply:
The whole point is to not use a fresh version of the component and to reuse the dismounted one again.

Reply:
As an example: We have a "table grid" component that has some filter fields and that component asks the backend for the data if is it not already present. This component is part of a tab group and when I switch to another tab I can simply dismount it and replace it with other content. The other one may or may not be also a tablegrid component and should not "overwrite" the first one, but replace it. So the first one gets dismounted but keeps state. Any running Goroutines can be stopped because of the "OnDismount". When it gets mounted later again, all state is still present.

I am interested in what you mean by "your implementation would me more safe...". What is the unsafe part in your experience?

Reply:
> I am interested in what you mean by "your implementation would me more safe...". What is the unsafe part in your experience?

I personally ran into issues when keeping reference on a component that is multiple times mounted/dismounted. But to be honest, I may have fixed it since it is working for you. I don't really remember.

As for your example, you are using a `+=` to assign a title ```func (c *tab) OnInit() { c.Title+="+" }```:

1. `a.OnInit` => `"+"`
2. `a.OnMount` => `"+"`
3. `a.OnDismount` => `"+"`
4. `b` init/mount.dismount
5. `a.OnInit` => `"++"`

If I understand well, what you want to do with `OnInit` is that it has to be called just once at the very first mount.

The reasons I'm not feeling good with this are:
- `OnInit` purpose is to set some component fields before the first rendering (or mount). When a component is dismounted, it is considered dead/fresh. Reusing it is like creating a new instance. Skipping `OnInit` introduces an exception in the component lifecycle and adds complexity to the understanding of it, therefore could be also perceived as a bug.
- Having it called once encourages users to keep references on components, which is something I try to avoid since I got some issues with it and I'm not sure I was able to fix this properly. I should take the time to put my hand into it of course but it takes time and you gave me other important things to add that I want to get right ;).
- If this is the behavior you seek, you can hack something for your use case to get the expected behavior.
- I want to keep the API as thin as possible

Reply:
I have a quite demanding test application that has canvas drawing (using `requestAnimationFrame`), two kinds of sockets (simple communication and a subscription based chat), some rest API calls for big datasets, deeply layered components (like tab in tab, containing multiple canvases and chats and API filled tables), huge components (like putting multiple big layouts all into one page), running 10 canvases all at once, creating audio on the fly), multiple stateful components implementing a simple game and a very basic spreadsheet. I want to figure out how stable and performant all of this is. Currently, I am rewriting it to not use `app.If` and using the "OnInit()" on multiple occasions. This all seems to work quite well currently. For what I do there, the "OnInit()" makes stuff much easier and understandable. I felt it very (very!) odd that "Render" is called before "OnMount" all the time. But maybe I have a different mindset about how this should work. I find it very normal wanting to have "something" which can mount and dismount without being recreated all the time.

Reply:
I see the point for mounting called after the first render. Maybe I should take a look back on this and make it happen. This would avoid getting and OnInit interface. 

For the behavior your seek, you could solve this with a sync.Once (see the edit above), whether it is in a modifed OnMount, or in a OnInit.

I’ll try to change the OnMount behavior. I like this.

Reply:
Well, it works perfectly fine as I made it in this PR. I just finished rewriting and testing. I could send you a link to check it out (but not public).

Will Render not be called on the server in the PreRender case?